### PR TITLE
feat: support contextual tuples in Check requests that are optimized

### PR DESCRIPTION
**Pull request by miparnisari**
## Description

- Added a new `ReadStartingWithUserOptions.WithResultsSortedAscending`
   - For now, this will only be read within storage wrappers, not within datastores themselves.
- In Check API fast path v2:
   - use this option to create a combination of sorted iterators.
   - enable the fast path v2 optimization when `len(req.contextual tuples) > 0`

## Dependencies

https://github.com/openfga/openfga/pull/2197

### CODE REVIEW

File: internal/graph/check_test.go [hunk 1]
```diff
@@ -28,6 +28,7 @@ import (
 	"github.com/openfga/openfga/pkg/logger"
 	"github.com/openfga/openfga/pkg/storage"
 	"github.com/openfga/openfga/pkg/storage/memory"
+	"github.com/openfga/openfga/pkg/storage/storagewrappers"
```
Comment (miparnisari) [inline internal/graph/check_test.go:31]: @jpadilla going back to https://github.com/openfga/openfga/pull/2117#discussion_r1862797277

I would need to move your storage wrapper together with the other storage wrappers, otherwise the linter complains:

```
internal/graph/builder.go:1: : import cycle not allowed in test (typecheck)
package graph
```

File: pkg/storage/storagewrappers/combinedtuplereader.go [hunk 2]
```diff
@@ -16,15 +17,35 @@ func NewCombinedTupleReader(
...
 		RelationshipTupleReader: ds,
-		contextualTuples:        contextualTuples,
 	}
+
+	co := make([]*openfgav1.TupleKey, len(contextualTuples))
+	cu := make([]*openfgav1.TupleKey, len(contextualTuples))
+	for i, t := range contextualTuples {
+		cu[i] = tuple.NewTupleKeyWithCondition(t.GetObject(), t.GetRelation(), t.GetUser(), t.GetCondition().GetName(), t.GetCondition().GetContext())
+		co[i] = tuple.NewTupleKeyWithCondition(t.GetObject(), t.GetRelation(), t.GetUser(), t.GetCondition().GetName(), t.GetCondition().GetContext())
+	}
```
Comment (miparnisari) [inline pkg/storage/storagewrappers/combinedtuplereader.go:29]: FYI to avoid race conditions in tests 😭

File: pkg/storage/iterators.go [hunk 3]
```diff
@@ -397,3 +399,124 @@ func NewConditionsFilteredTupleKeyIterator(iter TupleKeyIterator, filter TupleKe
...
+	return func(t *openfgav1.Tuple) string {
+		return t.GetKey().GetUser()
+	}
+}
+
+func ObjectMapper() TupleMapper {
+	return func(t *openfgav1.Tuple) string {
+		return t.GetKey().GetObject()
+	}
+}
```
Comment (miparnisari) [inline pkg/storage/iterators.go:424]: TODO: these belong, conceptually, in `internal/graph/tuplemapper.go`, but i don't want to have `storage` import `graph`.

File: pkg/storage/storagewrappers/combinedtuplereader.go [hunk 4]
```diff
@@ -15,16 +16,28 @@ import (
 func NewCombinedTupleReader(
 	ds storage.RelationshipTupleReader,
 	contextualTuples []*openfgav1.TupleKey,
-) storage.RelationshipTupleReader {
-	return &CombinedTupleReader{
+) *CombinedTupleReader {
```
Comment (miparnisari) [inline pkg/storage/storagewrappers/combinedtuplereader.go:19]: FYI so i can make assertions on the inner fields in tests. Also it's a good pattern to return structs instead of interfaces so that the caller can decide which interface it needs. https://medium.com/@cep21/preemptive-interface-anti-pattern-in-go-54c18ac0668a

File: internal/graph/check_fast_path.go [hunk 5]
```diff
@@ -928,7 +928,7 @@ func (c *LocalChecker) recursiveFastPath(ctx context.Context, req *ResolveCheckR
 
 	userIter, err := checkutil.IteratorReadStartingFromUser(ctx, typesys, ds, req,
 		tuple.ToObjectRelationString(tuple.GetType(req.GetTupleKey().GetObject()), req.GetTupleKey().GetRelation()),
-		nil)
+		nil, true)
```
Comment (elbuo8) [inline internal/graph/check_fast_path.go:931]: Why? This still relies in hashsets.

File: pkg/storage/iterators.go [hunk 6]
```diff
@@ -411,3 +412,121 @@ func NewConditionsFilteredTupleKeyIterator(iter TupleKeyIterator, filter TupleKe
...
+
+// NewOrderedCombinedIterator is a thread-safe iterator that combines a list of iterators into a single ordered iterator.
+// All the input iterators must be individually ordered already according to mapper.
+// Iterators can yield the same value (as defined by mapper) multiple times, but it will only be returned once.
+func NewOrderedCombinedIterator(mapper TupleMapper, sortedIters ...TupleIterator) *OrderedCombinedIterator {
+	pending := make([]TupleIterator, 0, len(sortedIters))
+	for _, sortedIter := range sortedIters {
+		if sortedIter != nil {
```
Comment (elbuo8) [inline pkg/storage/iterators.go:445]: In what case we would pass `nil` iterators? I believe I've seen this elsewhere, doesn't need to be handled in this PR.

File: pkg/storage/storagewrappers/combinedtuplereader.go [hunk 7]
```diff
@@ -15,16 +16,28 @@ import (
...
+
+	ctr.contextualTuplesOrderedByObjectID = cu
+
+	return ctr
 }
 
 type CombinedTupleReader struct {
 	storage.RelationshipTupleReader
-	contextualTuples []*openfgav1.TupleKey
+	contextualTuplesOrderedByObjectID []*openfgav1.TupleKey
```
Comment (elbuo8) [inline pkg/storage/storagewrappers/combinedtuplereader.go:40]: Why does this need to change?

File: pkg/storage/iterators.go (hunk 6 above)
Comment (miparnisari) [reply pkg/storage/iterators.go:445]: In no case. I copied from the other iterators

File: pkg/storage/iterators_test.go [hunk 8]
```diff
@@ -318,6 +318,141 @@ func TestCombinedIterator(t *testing.T) {
...
+		require.ErrorIs(t, err, ErrIteratorDone)
+	})
+
+	t.Run("Next", func(t *testing.T) {
+		var testcases = map[string]struct {
+			iter1    TupleIterator
+			iter2    TupleIterator
+			expected []*openfgav1.Tuple
+		}{
+			`removes_duplicates`: {
```
Comment (miparnisari) [inline pkg/storage/iterators_test.go:339]: missing test: when there are duplicates within one iterator

File: pkg/storage/iterators.go (hunk 6 above)
Comment (elbuo8) [reply pkg/storage/iterators.go:449]: Can we clean that up in a subsequent PR across the others 🤔

File: pkg/storage/iterators_test.go [hunk 9]
```diff
@@ -318,6 +318,204 @@ func TestCombinedIterator(t *testing.T) {
...
+				}),
+				expected: []*openfgav1.Tuple{
+					{Key: tuple.NewTupleKey("document:1", "2", "user:a")},
+					{Key: tuple.NewTupleKey("document:1", "2", "user:b")},
+					{Key: tuple.NewTupleKey("document:1", "2", "user:c")},
+					{Key: tuple.NewTupleKey("document:1", "2", "user:d")},
+					{Key: tuple.NewTupleKey("document:1", "2", "user:e")},
+					{Key: tuple.NewTupleKey("document:1", "2", "user:f")},
+				},
+			},
```
Comment (adriantam) [inline pkg/storage/iterators_test.go:384]: Can we add test where the last entry is also duplicate?

File: internal/graph/check_fast_path.go [hunk 10]
```diff
@@ -928,7 +928,7 @@ func (c *LocalChecker) recursiveFastPath(ctx context.Context, req *ResolveCheckR
 
 	userIter, err := checkutil.IteratorReadStartingFromUser(ctx, typesys, ds, req,
 		tuple.ToObjectRelationString(tuple.GetType(req.GetTupleKey().GetObject()), req.GetTupleKey().GetRelation()),
-		nil)
+		nil, false)
```
Comment (adriantam) [inline internal/graph/check_fast_path.go:931]: A bit concern that in some cases the iterator read starting from user is not sorted - would it lead to error if the caller relies on it in some path.

Comment (miparnisari) [reply internal/graph/check_fast_path.go:931]: This is funny. @elbuo8 told me that putting `true` here was unnecessary. https://github.com/openfga/openfga/pull/2150#discussion_r1892402696

You both have a point. I guess I'll add a godoc explaining why this is false

File: pkg/storage/iterators_test.go (hunk 9 above)
Comment (miparnisari) [reply pkg/storage/iterators_test.go:397]: Added removes_duplicates_across_iterators_last_entry, let me know if it's what you had in mind

File: pkg/storage/iterators.go [hunk 11]
```diff
@@ -411,3 +415,146 @@ func NewConditionsFilteredTupleKeyIterator(iter TupleKeyIterator, filter TupleKe
...
+	once        *sync.Once
+	mapper      TupleMapper
+	pending     []TupleIterator  // GUARDED_BY(mu)
+	lastHead    *openfgav1.Tuple // GUARDED_BY(mu)
+	lastYielded *openfgav1.Tuple // GUARDED_BY(mu)
+}
+
+var _ TupleIterator = (*OrderedCombinedIterator)(nil)
+
+// TODO(miparnisari): this interface & impls belong, conceptually, in internal/graph/tuplemapper.go, but i don't want to have storage import graph.
```
Comment (elbuo8) [inline pkg/storage/iterators.go:430]: Where is `graph` being used here?

These already exist, why are we re-declaring them?

File: pkg/storage/iterators.go [hunk 12]
```diff
@@ -411,3 +415,146 @@ func NewConditionsFilteredTupleKeyIterator(iter TupleKeyIterator, filter TupleKe
...
+		head, err := iter.Head(ctx)
+		if err != nil {
+			if errors.Is(err, ErrIteratorDone) {
+				iter.Stop()
+				c.pending[pendingIdx] = nil
+				continue
+			}
+			return minIdx, err
+		}
+
```
Comment (elbuo8) [inline pkg/storage/iterators.go:512]: Why not store `c.mapper(head)` if it will be called in all 3 code paths.

Same applies for ` c.mapper(c.lastYielded)` within the `c.lastYielded != nil` block.

File: pkg/storage/iterators.go (hunk 11 above)
Comment (miparnisari) [reply pkg/storage/iterators.go]: Definitely

File: pkg/storage/iterators.go (hunk 12 above)
Comment (miparnisari) [reply pkg/storage/iterators.go]: I moved these changes to a different PR and it got merged already. If you feel strongly about this I can definitely send a follow-up PR.

File: pkg/storage/storagewrappers/combinedtuplereader.go (hunk 7 above)
Comment (senojj) [reply pkg/storage/storagewrappers/combinedtuplereader.go:40]: perhaps could instead be a doc comment on the property?

File: pkg/storage/storagewrappers/combinedtuplereader.go (hunk 4 above)
Comment (senojj) [reply pkg/storage/storagewrappers/combinedtuplereader.go:19]: I'd like to go a step further and ask "why return a pointer to this struct?". The struct contains two pointers, for a total of 16 bytes, so copying an instance of this type is likely less expensive than the inevitable escape to the heap and cost of reading the value from the heap.

**Review by elbuo8 — APPROVED** [review]

File: internal/graph/check_test.go (hunk 1 above)
Comment (miparnisari) [reply internal/graph/check_test.go]: https://github.com/openfga/openfga/pull/2219

File: pkg/storage/storagewrappers/combinedtuplereader.go (hunk 4 above)
Comment (miparnisari) [reply pkg/storage/storagewrappers/combinedtuplereader.go:19]: This is a good idea. I can send a refactor PR.

File: internal/graph/check_fast_path.go (hunk 10 above)
Comment (miparnisari) [reply internal/graph/check_fast_path.go:933]: Updated existing tests.

File: pkg/storage/iterators.go (hunk 11 above)
Comment (miparnisari) [reply pkg/storage/iterators.go]: https://github.com/openfga/openfga/pull/2226

**Review by elbuo8 — APPROVED** [review]

File: pkg/storage/storagewrappers/combinedtuplereader.go (hunk 2 above)
Comment (miparnisari) [reply pkg/storage/storagewrappers/combinedtuplereader.go]: Partly tracked in https://github.com/openfga/openfga/issues/925
